yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+    logInfo(s"Creating domain $domainId with" +
+      s" readers: readers  and writers: $writers")
+
+    // create the timeline domain with the reader and writer permissions
+    val timelineDomain = new TimelineDomain()
+    timelineDomain.setId(domainId)
+    timelineDomain.setReaders(readers)
+    timelineDomain.setWriters(writers)
+    try {
+      getTimelineClient.putDomain(timelineDomain)
+    } catch {
+      case e: Exception => {
+        logError("cannot create the domain")
+        // fallback to default
+        domainId = null
+      }
+    }
+    domainId
+  }
+
+  /**
+   * Start the service, calling the service's `init()` and `start()` actions in the
+   * correct order
+   * @param context spark context
+   * @param appId YARN application ID
+   * @param attemptId YARN attempt ID if known. Otherwise, `None`.
+   * @return true if the service is hooked up to the timeline service; that is: it is live
+   */
+  def start(context: SparkContext,
+      appId: ApplicationId,
+      attemptId: Option[ApplicationAttemptId]): Boolean = {
+    require(context != null, "Null context parameter")
+    bindToYarnApplication(appId, attemptId)
+    
+    val yarnConf = new YarnConfiguration(context.hadoopConfiguration)
+    // the init() operation checks the state machine & prevents invocation out of sequence
+    init(yarnConf)
+    this.sparkContext = context
+
+    // work out the attempt ID from the YARN attempt ID. No attempt, assume "1".
+    // this is assumed by the AM, which uses it when creating a path to an attempt
+    val attempt1 = attemptId match {
+      case Some(attempt) => attempt.getAttemptId.toString
+      case None => CLIENT_BACKEND_ATTEMPT_ID
+    }
+    setContextAppAndAttemptInfo(Some(appId.toString), Some(attempt1))
+    batchSize = sparkContext.conf.getInt(BATCH_SIZE, batchSize)
+
+    start()
+    if (timelineServiceEnabled) {
+      true
+    } else {
+      logInfo("Yarn timeline service not available, disabling client.")
+      false
+    }
+  }
+
+  /**
+   * Service start.
+   *
+   * If the timeline client is enabled,
+   * Create the timeline client, the timeline domain, and the event handling thread.
+   *
+   * Irrespective of the timeline enabled flag, the service will attempt to register
+   * as a listener for events. They will merely be discarded.
+   */
+  override protected def serviceStart {
+    require(sparkContext != null, "No spark context set")
+    val conf: Configuration = getConfig
+    if (timelineServiceEnabled) {
+      timelineWebappAddress = getTimelineEndpoint(conf)
+      logInfo(s"Starting $this")
+      logInfo(s"Spark events will be published to the Timeline service at $timelineWebappAddress")
+      timelineClient = Some(createTimelineClient)
+      domainId = createTimelineDomain
+      // declare that the processing is started
+      stopped.set(false)
+      eventHandlingThread = Some(new Thread(new Dequeue(), "HistoryEventHandlingThread"))
```

Comment:
It ensures that events listened to are pushed out to the timeline service on a separate thread, so if the endpoint is (temporarily) unreachable then the event reporter thread isn't blocked. It also allows for some time for events to be pushed out at the end of the run and yet still stop the service after a delay (we don't want AM teardown to block because the timeline endpoint is playing up).

The alternative would be for synchronous publishing in the event listener, with the shutdown logic spinning something off there and then. In some respects it would actually make testing easier —but what would the impact on the rest of the spark listeners be if it was blocking?